### PR TITLE
Add Muzzle reference creation for interfaces

### PR DESCRIPTION
**Pull request by tylerbenson**
This will allow muzzle to fail if an implemented interface is missing.  Additional work should be done in the future to ensure the implementation satisfies the interfaces and abstract types. (https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/1357)

### CODE REVIEW

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/muzzle/Reference.java [hunk 1]
```diff
@@ -129,11 +131,6 @@ public Reference merge(final Reference anotherReference) {
     return merged;
   }
 
-  @Override
```
Comment (richardstartin) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/muzzle/Reference.java:132]: Seems like an unrelated change and a net negative.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/muzzle/Reference.java [hunk 2]
```diff
@@ -9,10 +9,12 @@
 import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Set;
+import lombok.ToString;
 import net.bytebuddy.jar.asm.Opcodes;
 import net.bytebuddy.jar.asm.Type;
 
 /** An immutable reference to a jvm class. */
+@ToString
```
Comment (richardstartin) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/muzzle/Reference.java:17]: Doesn't produce the same result as before. In fact, I can't see and therefore reason about what it does produce any more. We have an annotation from a library which messes with IDE integrations so we can avoid a one liner.

Comment (tylerbenson) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/muzzle/Reference.java:17]: I can revert the lombok change, but it was useful when trying to debug this problem as the previous `toString` only printed the name.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/muzzle/ReferenceCreator.java [hunk 3]
```diff
@@ -180,9 +180,16 @@ public void visit(
...
     refSourceTypeInternalName = refSourceType.getInternalName();
-    // Additional references we could check
-    // - supertype of class and visible from this package
-    // - interfaces of class and visible from this package
+
+    // Add references to each of the interfaces.
+    for (String iface : interfaces) {
+      addReference(
+          new Reference.Builder(iface)
+              .withSource(refSourceClassName, 0) // We don't have a specific line number to use.
```
Comment (richardstartin) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/muzzle/ReferenceCreator.java:188]: this is inconsistent with behaviour elsewhere in the class where the value -1 is used when no line number is known.

File: dd-java-agent/testing/src/test/java/muzzle/TestClasses.java [hunk 4]
```diff
@@ -95,10 +95,9 @@ public static boolean instanceofMethod(final Object a) {
 
   // Can't test this until java 7 is dropped.
   public static class InDyAdvice {
-    //    public static MethodBodyAdvice.SomeInterface indyMethod(
-    //        final MethodBodyAdvice.SomeImplementation a) {
+    //    public static MethodBodyAdvice.HasMethod indyMethod(final MethodBodyAdvice.HasMethod a) {
     //      Runnable aStaticMethod = MethodBodyAdvice.B::aStaticMethod;
-    //      return a::someMethod;
+    //      return a::requiredMethod;
```
Comment (tylerbenson) [inline dd-java-agent/testing/src/test/java/muzzle/TestClasses.java:100]: This is also an unrelated change... I uncommented it to verify it still worked and found it didn't, so I updated it here.

File: dd-java-agent/testing/src/test/groovy/muzzle/ReferenceCreatorTest.groovy [hunk 5]
```diff
@@ -49,6 +49,19 @@ class ReferenceCreatorTest extends AgentTestRunner {
...
+    def a = references.get('muzzle.TestClasses$MethodBodyAdvice$A')
+    def b = references.get('muzzle.TestClasses$MethodBodyAdvice$B')
+
+    expect:
+    references.keySet() == ['muzzle.TestClasses$MethodBodyAdvice$A', 'muzzle.TestClasses$MethodBodyAdvice$B'].toSet()
+
+    and: "doesn't need to reference self"
+    !(a.getSources().any { it.name == 'muzzle.TestClasses$MethodBodyAdvice$A' })
+  }
+
```
Comment (tylerbenson) [inline dd-java-agent/testing/src/test/groovy/muzzle/ReferenceCreatorTest.groovy:64]: I didn't mean to check this test in... I'll remove it.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/muzzle/ReferenceCreator.java (hunk 3 above)
Comment (richardstartin) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/muzzle/ReferenceCreator.java]: I'm not sure that the existing behaviour is correct. Line numbers will be missing if code has been passed through an obfuscator, for instance, in which case `MethodVisitor.visitLineNumber` wouldn't be entered, and line number 0 may be a better fallback value than -1.

Comment (tylerbenson) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/muzzle/ReferenceCreator.java]: This only matters for our own code, so the case of an obfuscator doesn't likely apply.  I don't care between `-1` and `0`, but I did extract it out to a static final.

**Comment by tylerbenson** [comment]
I added this so that #2054 is able to distinguish a version where the implemented interface name changed/moved.

**Review by richardstartin — APPROVED** [review]

**Review by bantonsson — APPROVED** [review]

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java [hunk 6]
```diff
@@ -240,7 +240,8 @@ private void decrementRefAndMaybeWrite(boolean isRootSpan) {
       }
     }
     if (log.isDebugEnabled()) {
-      log.debug("t_id={} -> expired reference. pending count={}", traceId, count);
```
Comment (richardstartin) [inline dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java:243]: Why was this committed with this PR?

**Comment by richardstartin** [comment]
@tylerbenson this PR is called "Add Muzzle reference creation for interfaces" yet a change to `PendingTrace` has been brought in with it after review? Please change the name of the PR and its description to better describe the change.

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java (hunk 6 above)
Comment (dougqh) [reply dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java:243]: Yes, agreed.  That seems completely unrelated and should not be in this PR.

**Review by dougqh — COMMENTED** [review]
The unrelated change to PendingTrace needs to be removed.  Or needs justification.

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java (hunk 6 above)
Comment (tylerbenson) [reply dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java:243]: it is unrelated to the pr.  I made this change to troubleshoot some build failures, but the build passed following this change.  It seems like a useful and very minor change so I left it in.

**Comment by tylerbenson** [comment]
For reference you can see the history of me running the test multiple times to try to figure out what was wrong here: https://app.circleci.com/pipelines/github/DataDog/dd-trace-java?branch=tyler%2Fmuzzle-super
